### PR TITLE
Fix some code quality issues

### PR DESCRIPTION
**Pull request by srijan-deepsource**
This pull request fixes some of the quality issues raised by DeepSource on my fork of this repository.

Summary of fixes:
- Remove redundant `None` default
- Use identity test to compare types
- Pass string format arguments as logging method parameters
- Refactor the comparison involving `not`
- Use literal syntax instead of function calls to create a data structure
- Refactor `if` expression
- Iterate dictionary directly
- Use `with` statement to open file
- Merge consecutive isinstance calls

You can see all the issues raised by DeepSource [here](https://deepsource.io/gh/srijan-deepsource/readthedocs.org/issues/?category=recommended).
Some of the issues were automatically fixed by DeepSource, while I fixed others manually.

### Using DeepSource to continuously analyze your repository
1. Merge this PR. I have included a `.deepsource.toml` in this PR, which you can use to configure your analysis settings.
2. Install DeepSource on your repository [here](https://deepsource.io/signup).
3. Activate analysis [here](https://deepsource.io/gh/readthedocs/readthedocs.org/).

If you do not want to use DeepSource to continuously analyze this repo, I'll remove the `.deepsource.toml` from this PR and you can merge the rest of the fixes.

### CODE REVIEW

File: readthedocs/doc_builder/base.py [hunk 1]
```diff
@@ -134,8 +134,9 @@ def create_index(self, extension='md', **__):
 familiar with Read the Docs.
                 """
 
-                index_file.write(index_text.format(dir=docs_dir, ext=extension))
-                index_file.close()
+                with open(index_filename, 'w+') as index_file:
```
Comment (stsewd) [inline readthedocs/doc_builder/base.py:137]: you should remove the open call at line ~120.

**Review by stsewd — COMMENTED** [review]
Thanks, not sure about using deepsource. Can you add that in a different PR?

File: readthedocs/doc_builder/base.py (hunk 1 above)
Comment (srijan-deepsource) [reply readthedocs/doc_builder/base.py:137]: Right. I missed that.

**Comment by srijan-deepsource** [comment]
> Thanks, not sure about using deepsource. Can you add that in a different PR?

Done. Opened #7495

**Review by stsewd — APPROVED** [review]
Thanks!

**Comment by stsewd** [comment]
Looks like there are some tests that need to be updated. 

```diff
diff --git a/readthedocs/rtd_tests/tests/test_build_notifications.py b/readthedocs/rtd_tests/tests/test_build_notificati
ons.py
index f09bdffc9..a92131f48 100644
--- a/readthedocs/rtd_tests/tests/test_build_notifications.py
+++ b/readthedocs/rtd_tests/tests/test_build_notifications.py
@@ -24,7 +24,10 @@ class BuildNotificationsTests(TestCase):
     def test_send_notification_none_if_wrong_version_pk(self, mock_logger):
         self.assertFalse(Version.objects.filter(pk=345343).exists())
         send_notifications(version_pk=345343, build_pk=None)
-        mock_logger.warning.assert_called_with("Version not found for given kwargs. {'pk': 345343}")
+        mock_logger.warning.assert_called_with(
+            'Version not found for given kwargs. %s',
+            {'pk': 345343},
+        )


     def test_send_notification_none(self):
diff --git a/readthedocs/rtd_tests/tests/test_celery.py b/readthedocs/rtd_tests/tests/test_celery.py
index aae93822b..90b28ba7e 100644
--- a/readthedocs/rtd_tests/tests/test_celery.py
+++ b/readthedocs/rtd_tests/tests/test_celery.py
@@ -334,7 +334,10 @@ class TestCeleryBuilding(TestCase):
             search_ranking={},
             search_ignore=[],
         )
-        mock_logger.warning.assert_called_with("Version not found for given kwargs. {'pk': 345343}")
+        mock_logger.warning.assert_called_with(
+            'Version not found for given kwargs. %s',
+            {'pk': 345343},
+        )

```

**Pull request merged**